jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostEstimator.scala

```diff
@@ -0,0 +1,622 @@
+   * If so, throw an exception unless this safety measure has been explicitly overridden
+   * via conf `xgboost.spark.ignoreSsl`.
+   */
+  private def validateSparkSslConf(spark: SparkSession): Unit = {
+
+    val sparkSslEnabled = spark.conf.getOption("spark.ssl.enabled").getOrElse("false").toBoolean
+    val xgbIgnoreSsl = spark.conf.getOption("xgboost.spark.ignoreSsl").getOrElse("false").toBoolean
+
+    if (sparkSslEnabled) {
+      if (xgbIgnoreSsl) {
+        logger.warn(s"spark-xgboost is being run without encrypting data in transit!  " +
+          s"Spark Conf spark.ssl.enabled=true was overridden with xgboost.spark.ignoreSsl=true.")
+      } else {
+        throw new Exception("xgboost-spark found spark.ssl.enabled=true to encrypt data " +
+          "in transit, but xgboost-spark sends non-encrypted data over the wire for efficiency. " +
+          "To override this protection and still use xgboost-spark at your own risk, " +
+          "you can set the SparkSession conf to use xgboost.spark.ignoreSsl=true.")
+      }
+    }
+  }
+
+  /**
+   * Validate the parameters before training, throw exception if possible
+   */
+  protected[spark] def validate(dataset: Dataset[_]): Unit = {
+    validateSparkSslConf(dataset.sparkSession)
+    val schema = dataset.schema
+    SparkUtils.checkNumericType(schema, $(labelCol))
+    if (isDefinedNonEmpty(weightCol)) {
+      SparkUtils.checkNumericType(schema, $(weightCol))
+    }
+
+    if (isDefinedNonEmpty(baseMarginCol)) {
+      SparkUtils.checkNumericType(schema, $(baseMarginCol))
+    }
+
+    val taskCpus = dataset.sparkSession.sparkContext.getConf.getInt("spark.task.cpus", 1)
+    if (isDefined(nthread)) {
+      require(getNthread <= taskCpus,
+        s"the nthread configuration ($getNthread) must be no larger than " +
+          s"spark.task.cpus ($taskCpus)")
+    } else {
+      setNthread(taskCpus)
+    }
+  }
+
+  def train(dataset: Dataset[_]): M = {
+    validate(dataset)
+
+    val rdd = if (isPluginEnabled(dataset)) {
+      getPlugin.get.buildRddWatches(this, dataset)
+    } else {
+      val (input, columnIndexes) = preprocess(dataset)
+      toRdd(input, columnIndexes)
+    }
+
+    val xgbParams = getXGBoostParams
+
+    val runtimeParams = getRuntimeParameters(dataset.sparkSession.sparkContext.isLocal)
+
+    val (booster, metrics) = XGBoost.train(rdd, runtimeParams, xgbParams)
+
+    val summary = XGBoostTrainingSummary(metrics)
+    copyValues(createModel(booster, summary))
+  }
+
+  override def copy(extra: ParamMap): Learner = defaultCopy(extra).asInstanceOf[Learner]
+}
+
+/**
+ * Indicate what to be predicted
+ *
+ * @param predLeaf    predicate leaf
+ * @param predContrib predicate contribution
+ * @param predRaw     predicate raw
+ * @param predTmp     predicate probability for classification, and raw for regression
+ */
+private[spark] case class PredictedColumns(
+    predLeaf: Boolean,
+    predContrib: Boolean,
```

Comment:
Does spark support shap feature interaction?

Reply:
As far as I know, the answer is no.